docs/adr/0007-seperate-binding-generators.md

```diff
@@ -0,0 +1,142 @@
+```
+
+There are a few specific ways to implement this
+
+## **Implementation Option 1: Have `uniffi_bindgen` shell out to `uniffi_bindgen_{language}`**
+Using this approach, we would recommend users to install both the `uniffi_bindgen` **binary** and the `uniffi_bindgen_kotlin` (using kotlin as an example) **binary**. Then, the user would run the following:
+1. The user runs `uniffi-bindgen generate -l kotlin -o <OUT_PATH> <UDL_PATH>`
+1. `uniffi-bindgen` gets its own version, let's say `0.16` as an example.
+1. `uniffi-bingen` shells out to `uniffi-bingen-kotlin -o <OUT_PATH> -v 0.16 <UDL_PATH>`
+1. `uniffi-bindgen-kotlin` asserts that it was compiled using the same `0.16` version, otherwise it panics.
+1. `uniffi-bindgen-kotlin` then calls a `generate_bindings` function in `uniffi-bindgen` (as a library dependency)
+1. `uniffi-bindgen`'s `generate_bindings` takes a generic parameter type that implements a trait `BindingGenerator`.
+1. `uniffi-bindgen` parses the `udl`, then calls the generic type's `write_bindings` function (the implementation lives in `uniffi-bingen-kotlin`)
+
+
+## **Implementation Option 2: Have `uniffi_bindgen` get dynamically linked with exactly one `uniffi_bindgen_{language}`**
+This approach uses the specific `uniffi_bindgen_{language}` as plugins. It essentially replaces the shelling out in `Implementation Option 1` with calling an exposed unsafe `C` API in a dynamically linked crate. 
+
+I'm not 100% sure how to let `uniffi_bindgen` choose which crate it gets dynamically linked with on run time, but it's possible.
+
+
+## **Implementation Option 3: Have each `uniffi_bindgen_{language}` be run directly**
+This approach has `uniffi_bindgen` only provide scaffolding and traits for each `uniffi_bindgen_{language}` to implement. `uniffi_bindgen` would also include some generic functions that can be called from `uniffi_bindgen_{language}` by treating `uniffi_bindgen` as a library dependency. The flow would look like this:
+
+1. A user would invoke `uniffi-bindgen-kotlin bindings -o <OUT_PATH> <UDL_PATH>`
+1. `uniffi-bindgen-kotlin` would call a generic function `generate_bindings` in `uniffi-bindgen` that takes type parameter `B: BindingGenerator`
+1. `uniffi-bindgen` parses the `udl`, then calls the generic type's `write_bindings` function (which lives in `uniffi-bingen-kotlin`)
+
+Version compatibility is a serious consideration with this approach, and we would benefit from exposing a `scaffolding` option to `uniffi-bindgen-kotlin`. This way the same `uniffi-bindgen-kotlin` is used to generate both scaffolding and bindings. We can also add a warning to `uniffi-bindgen`'s scaffolding that we recommend using a specific binary (like `uniffi-bindgen-kotlin`) to generate the scaffolding as well. Note: The scaffolding generating logic would still live in `uniffi-bindgen` we would only expose it to be called. `uniffi-bindgen-kotlin` would not modify the logic at all, this is purely for version compatibility.
```

Comment:
Scaffolding can happen in a crate's `build.rs`. That's foreign-language-independent and it has to stay so.
~~In Glean we generate a single library that could be consumed by different foreign languages, it doesn't make sense to encourage use of a specificy language binding cli for that.~~
Right now we don't need to think about how we build that, a simple `cargo build` will do the right job, as long as we use the same version for the bindings.
Additionally `uniffi` is _also_ a runtime dependency, so another version that must match.
With this we still need to ensure `uniffi-bindgen-{kotlin,swift,python}` are compatible, or make our builds more complex for when they aren't (choosing the right combo of dependency & tool).

Reply:
FWIW I quite like the way that we've protected the language-independent nature of the scaffolding, so I prefer the option where a top-level `uniffi-bindgen` continues to make that distinction clear.